### PR TITLE
feat!: add transaction execute and finalized time to vn+indexer+wallet

### PR DESCRIPTION
**Pull request by sdbondi**
Description
---
Add finalized_at to state store transactions 
Return the finalization and execution time to jrpc get transaction result
Show execution and finalization time in transactions page in VN web ui
fix(indexer): improve error output for submit_transaction
fix(transaction): exclude virtual substates from Transaction::to_referenced_substates
fix(autofill): scan for next version if substate DOWN
fix(tariswap-test): request transaction results via SDK + improve logs

Motivation and Context
---
Allows us to get the actual time to finalize the transaction

How Has This Been Tested?
---
Checked finalize time in web ui

What process can a PR reviewer use to test or verify this change?
---
Submit transactions and check the finalization time in VN web ui

Breaking Changes
---

- [ ] None
- [x] Requires data directory to be deleted
- [ ] Other - Please specify

### CODE REVIEW

File: applications/tari_validator_node_web_ui/src/utils/helpers.tsx [hunk 1]
```diff
@@ -89,26 +90,81 @@ function emptyRows(page: number, rowsPerPage: number, array: any[]) {
...
+}
+function displayDuration(duration: Duration) {
+  if (duration.secs === 0) {
+    if (duration.nanos > 1000000) {
+      return `${(duration.nanos / 1000000).toFixed(2)}ms`;
+    }
+    if (duration.nanos > 1000) {
+      return `${(duration.nanos / 1000).toFixed(2)}µs`;
+    }
+    return `${duration.nanos / 1000}ns`;
```
Comment (Cifko) [inline applications/tari_validator_node_web_ui/src/utils/helpers.tsx]:
```suggestion
    return `${duration.nanos}ns`;
```

File: applications/tari_validator_node_web_ui/src/utils/helpers.tsx [hunk 2]
```diff
@@ -89,26 +90,81 @@ function emptyRows(page: number, rowsPerPage: number, array: any[]) {
...
+  }
+  if (duration.secs > 60 * 60) {
```
Comment (Cifko) [inline applications/tari_validator_node_web_ui/src/utils/helpers.tsx]:
```suggestion
  if (duration.secs >= 60 * 60) {
```

File: applications/tari_validator_node_web_ui/src/utils/helpers.tsx [hunk 3]
```diff
@@ -89,26 +90,81 @@ function emptyRows(page: number, rowsPerPage: number, array: any[]) {
...
+    return `${(duration.secs / 60 / 60).toFixed(0)}h${(
+      duration.secs / 60
+    ).toFixed(0)}m`;
+  }
+  if (duration.secs > 60) {
```
Comment (Cifko) [inline applications/tari_validator_node_web_ui/src/utils/helpers.tsx]:
```suggestion
  if (duration.secs >= 60) {
```

File: applications/tari_validator_node_web_ui/src/utils/helpers.tsx [hunk 4]
```diff
@@ -89,26 +90,81 @@ function emptyRows(page: number, rowsPerPage: number, array: any[]) {
...
+    return `${(duration.secs / 60).toFixed(0)}m${(duration.secs % 60).toFixed(
+      0,
+    )}s`;
+  }
+  return `${duration.secs}.${(duration.nanos / 1000000).toFixed(0)}s`;
```
Comment (Cifko) [inline applications/tari_validator_node_web_ui/src/utils/helpers.tsx]: We need to pad it from the front so 1ms and 100ms will not both end up as 0.1s. Also remove trailing zeros.
```suggestion
  return `${duration.secs}.${(duration.nanos / 1000000).toFixed(0).padStart(3,0).replace(/0+$/,'')}s`;
```

**Review by Cifko — DISMISSED** [review]
some minor formatting comments

**Review by stringhandler — APPROVED** [review]

**Pull request merged**